crypto/rand/rand_lib.c

```diff
@@ -211,27 +194,357 @@ void rand_cleanup_int(void)
+}
+
+/*
+ * Return the |pool|'s buffer length to the caller.
+ */
+size_t RAND_POOL_length(RAND_POOL *pool)
+{
+    return pool->len;
+}
+
+/*
+ * Detach the |pool| buffer and return it to the caller.
+ * It's the responsibility of the caller to free the buffer
+ * using OPENSSL_secure_free().
+ */
+unsigned char *RAND_POOL_detach(RAND_POOL *pool)
+{
+    unsigned char * ret = pool->buffer;
+    pool->buffer = NULL;
+    return ret;
+}
+
+
+/*
+ * If every byte of the input contains |entropy_per_bytes| bits of entropy,
+ * how many bytes does one need to obtain at least |bits| bits of entropy?
+ */
+#define ENTROPY_TO_BYTES(bits, entropy_per_bytes) \
+    (((bits) + ((entropy_per_bytes) - 1))/(entropy_per_bytes))
+
+
+/*
+ * Checks whether the |pool|'s entropy is available to the caller.
+ * This is the case when entropy count and buffer length are high enough.
+ * Returns
+ *
+ *  |entropy|  if the entropy count and buffer size is large enough
+ *      0      otherwise
+ *     -1      on error 
+ */
+int RAND_POOL_entropy_available(RAND_POOL *pool)
+{
+    if (pool == NULL)
+        return -1;
+
+    if (pool->entropy < pool->requested_entropy)
+        return 0;
+
+    if (pool->len < pool->min_len)
+        return 0;
+
+    return pool->entropy;
+}
+
+/*
+ * Returns the (remaining) amount of entropy needed to fill
+ * the random pool.
+ *
+ * If an error occurs, -1 is returned.
+ */
+
+int RAND_POOL_entropy_needed(RAND_POOL *pool)
+{
+    if (pool == NULL)
+        return -1;
+
+    if (pool->entropy < pool->requested_entropy)
+        return pool->requested_entropy - pool->entropy;
+    
+    return 0;
+}
+
+/*
+ * Returns the number of bytes needed to fill the pool, assuming
+ * the input has 'entropy_per_byte' entropy bits per byte.
+ *
+ * If an error occurs, -1 is returned.
+ */
+
+int RAND_POOL_bytes_needed(RAND_POOL *pool, unsigned int entropy_per_byte)
+{
+    unsigned int bytes_needed;
+    
+    int entropy_needed = RAND_POOL_entropy_needed(pool);
+        
+    if (entropy_needed < 0 || entropy_per_byte <= 0)
+        return -1;
+
+    bytes_needed = ENTROPY_TO_BYTES((unsigned int)entropy_needed, entropy_per_byte);
+
+    if (bytes_needed > pool->max_len - pool->len)
+        return -1; /* not enough space left */
+    else if (bytes_needed < pool->min_len - pool->len)
+        bytes_needed = pool->min_len - pool->len; /* to meet the min_len requirement */
+    
+    return bytes_needed;
+}
+
+/*
+ * Returns the remaining number of bytes available
+ *
+ * If an error occurs, -1 is returned.
+ */
+
+int RAND_POOL_bytes_remaining(RAND_POOL *pool)
+{
+    if (pool == NULL)
+        return -1;
+    return pool->max_len - pool->len;
+}
+
+/*
+ * Add the contents of the |buffer| to the random pool.
+ * Return available amount of entropy after this operation.
+ * (see RAND_POOL_entropy_available(pool))
+ */
+int RAND_POOL_add(RAND_POOL *pool, const void *buffer, int num, double randomness)
+{
+    size_t len  = (size_t)num;
```

Comment:
Is there a reason not to make num a size_t?

Reply:
The signature is identical to the `add()` callback of the `RAND_METHOD`, because all the  entropy gathering callbacks used to call `RAND_add()`, which has been replaced by `RAND_POOL_add()`. (Interestingly `RAND_add()` ignores the return value of the `add()` method and returns void.)
```C
    int (*add) (const void *buf, int num, double randomness);
```
I kept the signature because I didn't want to mess around with the entropy gathering callbacks too much. I agree with you that the signature should be changed, also the `double randomness` needs to be replaced. But I think it is better to keep that for a followup PR which redesigns the entropy gathering callbacks (to make them more flexible and configurable).